### PR TITLE
io: Do not turn on automatically MustClean for TDirectory

### PR DESCRIPTION
**Pull request by pcanal**
Fix https://github.com/root-project/root/issues/13449

### CODE REVIEW

File: core/base/src/TDirectory.cxx [hunk 1]
```diff
@@ -213,7 +213,10 @@ void TDirectory::Append(TObject *obj, Bool_t replace /* = kFALSE */)
    }
 
    fList->Add(obj);
-   obj->SetBit(kMustCleanup);
+   // A priori TDirectory will not be 'shared owned' or referenced.
+   // If they are, let's really on the user to update the bit.
```
Comment (jalopezg-git) [inline core/base/src/TDirectory.cxx:217]:
```suggestion
   // A priori, a `TDirectory` object is assumed to not have shared ownership.
   // If it is, let's rely on the user to update the bit.
```

**Review by jalopezg-git — COMMENTED** [review]
I ignore why this was originally making the cleanup slow; I have a suggestion on the code comment though :slightly_smiling_face:!

File: core/base/src/TDirectory.cxx (hunk 1 above)
Comment (pcanal) [reply core/base/src/TDirectory.cxx:217]: > I ignore why this was originally making the cleanup slow 

The 'kMustCleanup' flags tells the `TObject` destructor to inform the `gROOT->GetListOfCleanups()` (via a call to `RecursiveRemove` that the object is being deleted.   `TFile` are on that list by default and the call to `RecursiveRemove` will (of course) recursively call `RecursiveRemove`.  The `TDirectory` then have to lookup in their list if the deleted object is there.  So in this case there is O(N^2) behavior/cost to do ... exactly nothing ...

**Review by jalopezg-git — APPROVED** [review]
LGTM!  However, I do have a couple of concerns:
- Does this change affect cleanup of objects owned by `gDirectory`?
- Maybe we could extend the code comment with some information from https://github.com/root-project/root/pull/13451#discussion_r1293968701.

**Comment by jalopezg-git** [comment]
@pcanal I saw the `ClingLoadUnloadFile.ConcurrentLoadUnloadOneLibPerThread` test failing on the new CI for this PR, on fedora38 and alma8; could you check?

**Comment by Axel-Naumann** [comment]
That failure is unrelated / preexisting (and it might benefit from your brain's input on why it fails with `bad_alloc` - I couldn't find any reason yet).

**Comment by pcanal** [comment]
> Does this change affect cleanup of objects owned by gDirectory?

No.  It just prevent the `TDirectory` itself to broadcast its own deletion.

**Pull request merged**